Polish and tests for the improved CachedFilesMixin

idan opened the pull request.

idan committed:
Improve docstring for CachedFilesMixin.post_process

Explain what "processing" actually means. Also fix typo in comment.

idan committed:
Use better terminology for two parts of post-processing

Files which match CachedFilesMixin.patterns are _adjusted_ with paths to the
hashed files.

idan commented:
I've given some thought to tests for this branch.

## The tests I'd like to have:
- test that non-adjustable files which exist on target are skipped when
  rerunning collectstatic
- test that adjustable files (anything that matches CachedFilesMixin.patterns)
  are always processed, even if their hash hasn't changed.

## Problems
- Testcases currently assume a static snapshot from one collectstatic run.
- Collectstatic running inside testcases are silent (verbosity 0). I don't have
  a way to inspect the output of collectstatic to see what was processed,
  skipped, etc.

To do these tests properly, I'd ideally have a way of getting the individual status for every file seen by collectstatic (copied/processed/skipped). Failing that, I could still test based on the summary counts (copied/symlinked,
unmodified, post-processed).

How should I best get a hold of the information in the context of tests?

My best thought was to move the logic from collectstatic's `handle_noargs()` into a separate method that returns (modified_files, unmodified_files, postprocessed_files). `handle_noargs()` would still be responsible for the output on [collectstatic.py:147](https://github.com/idangazit/django-staticfiles/blob/features/improved-cachedfilesmixin/staticfiles/management/commands/collectstatic.py#L147).

Thoughts?

jezdez commented:
@idangazit 

Sure, I could live with splitting up the command further if it helps specifically testing the command.

idan commented:
Forgot to mention: we could probably just write this as a new method in the existing `TestCollectionCachedStorage` testcase class. It'd probably look like this:

``` python

def test_post_processing(self):
    # run collectstatic "again", since it was run already in setUp()
    # check that project/documents/cached/css/img/window.png was skipped
    # check that project/documents/cached/css/window.css wasn't
```

jezdez commented:
Sounds good to me, too.

idan committed:
Fix typo

idan committed:
Fix pathname appended to post_processed_files

idan committed:
Split collectstatic logic into collect() method

Allows for easier calling from tests.

idan committed:
Test post_processing behavior

idan commented:
OK, this is functionally complete. All tests pass, one test added, one bug fixed. I can haz code review, plz?

jezdez commented on staticfiles/management/commands/collectstatic.py:
```diff
@@ -62,18 +62,20 @@ def __init__(self, *args, **kwargs):
             os.stat_float_times(False)
 
-    def handle_noargs(self, **options):
+    def _set_options(self, **options):
```
Ugh, yeah, let's not do private methods, could we do that in `__init__` instead?

idan replied:
Okay.

idan replied:
I don't think **init** isn't doable -- the options are coming in as arguments to handle_noargs, not available at **init** time, and I need to be able to pass the same options to collect().

jezdez replied:
Grr, right, awkward interface :/ Can we somehow go to using options directly by making sure the defaults of the command line options are correct set?

jezdez commented on staticfiles/tests/tests.py:
```diff
@@ -427,6 +429,31 @@ def test_template_tag_url(self):
+        }
+        from pprint import pprint
+
+        c = CollectstaticCommand()
```
One letter variables? Come on!

idan replied:
mea culpa.

jezdez commented on staticfiles/tests/tests.py:
```diff
@@ -427,6 +429,31 @@ def test_template_tag_url(self):
+            'use_default_ignore_patterns': True,
+            'ignore_patterns': ['*.ignoreme']
+        }
+        from pprint import pprint
```
Why importing that inline?

idan replied:
Whoops, holdover from debugging.

idan commented:
Will make the fixes in a couple of hours when the girls get to bed -- otherwise is this all sensible?

jezdez commented:
Yeah, LGTM otherwise. Thanks!

idan committed:
Remove debug code, improve variable name

idan committed:
Improve options handling in collectstatic

Instead of a private method which is called twice (once in handle_noargs and
once in collect), process options once in handle_noargs. Tests which call
collect() will have to manually call set_options() beforehand to configure
their desired options.

idan commented:
Ok, this seems like a cleaner solution.

It does away with the private method, and removes the need to double-set the options. Tests calling `collect` will have to call `set_option` on the command's class instance first, with the desired options. I think that's a reasonable approach, anyways we want to lock down a specific set of options for the tests.

I originally tried to do something involving the default options, but the way that `make_option` works, I can't directly read the default options outside of `handle_noargs`, where I get the options dict.
